server/container/lifecycle-api/src/main/java/org/apache/james/lifecycle/api/Disposable.java

```diff
@@ -29,4 +39,111 @@
      * Dispose the object
      */
     void dispose();
+
+    abstract class LeakAware implements Disposable {
+        public static class LeakDetectorException extends RuntimeException {
+            public LeakDetectorException() {
+                super();
+            }
+
+            public LeakDetectorException(String message) {
+                super(message);
+            }
+
+            public LeakDetectorException(String message, Throwable cause) {
+                super(message, cause);
+            }
+        }
+
+        public enum Level {
+            NONE,
+            SIMPLE,
+            ADVANCED,
+            TESTING;
+
+            static Level parse(String input) {
+                for (Level level : values()) {
+                    if (level.name().equalsIgnoreCase(input)) {
+                        return level;
+                    }
+                }
+                throw new IllegalArgumentException(String.format("Unknown level `%s`", input));
+            }
+        }
+
+        public static final ReferenceQueue<Object> REFERENCE_QUEUE = new ReferenceQueue<>();
+        public static final List<LeakAwareFinalizer> LEAK_REFERENCES = new ArrayList<>();
+        public static Level level;
+        private final AtomicBoolean isDisposed;
+
+        protected LeakAware() {
+            isDisposed = new AtomicBoolean(false);
+            level = Level.SIMPLE;    // todo
+            LEAK_REFERENCES.add(new LeakAwareFinalizer(this, REFERENCE_QUEUE));
+        }
+
+        public void disposed() {
+            isDisposed.set(true);
+        }
+
+        public static void track() {
+            if (leakDetectorEnabled()) {
+                Reference<?> referenceFromQueue;
+                while ((referenceFromQueue = REFERENCE_QUEUE.poll()) != null) {
+                    ((LeakAwareFinalizer) referenceFromQueue).detectLeak();
+                    referenceFromQueue.clear();
+                }
+            }
+        }
+
+        private static boolean leakDetectorEnabled() {
+            return level != Level.NONE;
+        }
+    }
+
+    class LeakAwareFinalizer extends PhantomReference<Object> {
+        private static final Logger LOGGER = LoggerFactory.getLogger(LeakAwareFinalizer.class);
+
+        private LeakAware referent;
+
+        public LeakAwareFinalizer(LeakAware referent, ReferenceQueue<? super Object> q) {
+            super(referent, q);
+//            this.referent = referent;
```

Comment:
I'm stuck here, 
`referent` is used by `isNotDisposed` method.
But if we set it, GC won't remove LeakAware from memory (REFERENCE_QUEUE always empty)

Reply:
That's the finalizer that get gced and that should thus not be referenced. That's the finalizer that should be in the queue.

From what I understood.

Reply:
https://docs.oracle.com/javase/8/docs/api/java/lang/ref/PhantomReference.html#get--

Use `PhantomReference::get` and don't store the reference yourself.

Reply:
`PhantomReference::get` always return NULL

Reply:
Ah ok I did not read the javadoc well enough

Reply:
The cleanup needs to be a callback to the reference cf https://dzone.com/articles/weak-soft-and-phantom-references-in-java-and-why-they-matter

Since Phantom doesn't have a link to the actual object, a typical pattern is to derive your own Reference type from Phantom and add some info useful for the final freeing, for example filename.

LeakAware likely need to provide a callback to clean its guts, of course without referencing itself.

Reply:
https://github.com/apache/commons-io/blob/aad6583bd2c18da84dd85bd61197c578febf388f/src/main/java/org/apache/commons/io/FileCleaningTracker.java#L84